### DISCUSSION TITLE
Which versioning should I use if I always want the latest docker image?

### DISCUSSION THREAD

**Discussion by paololazzari** [renovatebot/renovate#29653] [Request Help]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitHub, version 40.1.12

### Please tell us more about your question or problem

My docker images have a format like `0.1.0-c3b...`. 

I have tried `loose` versioning but it doesn't seem to care about the date in which the images were created, and in fact has created a PR for a non-latest version (presumably based on the git commit hash)

I want renovate to always get the **latest** image from the registry and use that in the PR. What versioning should I use?

**Comment by rarkins** [comment]
Do you mean that your image tags don't _sort_ from oldest to newest, so the only way to know the best one to use is to look at the publish date?

**Reply by paololazzari** [reply]
I mean that the version in the tag is not necessarily always incremented (I am aware this is not ideal), e.g.: `0.1.0-c3b` and `0.1.0-eac`. With `loose` versioning it created a PR using `0.1.0-eac` as the new image despite `0.1.0-c3b` being more recent.

**Reply by rarkins** [reply]
We don't currently have a "sort by date" functionality. Also note that not all registries make the release date easily accessible via an API call

**Reply by HonkingGoose** [reply]
## Maybe try SemVer versioning?

Did you consider following [the SemVer specification](https://semver.org/)? Appending commits to the end of versions is harder to understand and use than plain versions.

### Summary of SemVer rules

This table shows the progress of a project during it's lifetime according to SemVer:

| Commit content | SemVer version |
| --- | --- |
| start of project | `0.0.0` |
| new feature | `0.1.0` |
| whoops, needs a bugfix | `0.1.1` |
| new feature | `0.2.0` |
| new feature | `0.3.0` |
| new feature | `0.4.0` |
| bugfix for feature | `0.4.1` |
| Stable version | `1.0.0` |
| New feature | `1.1.0` |
| Breaking change, so new major version | `2.0.0` |

#### Once stable, breaking releases require new major version!

One important thing you'll see above, is that when you mark the project as stable with a `1.0.0` release, any breaking changes in the API/ABI of your program require a new major version.

#### Break stuff as often as you want in `0.x.x.` releases

As long as you are in the `0.x.x` range, you can make breaking changes in any release, according to SemVer.

### Benefits of SemVer

Using SemVer makes it easier for Renovate, and humans to find the correct actual update.

Another benefit of SemVer is that you can automate the release process and bumping the version with tools like [Semantic Release](https://github.com/semantic-release/semantic-release).

### How to migrate to SemVer versioning?

To migrate to SemVer you could create a new patch release: `0.1.1`, and then start following Semver.

#### Read the specs before you start!

Again, read the [the SemVer specification](https://semver.org/) before you start. The above is my best attempt at explaining this from memory. :wink: 

## Outro

I hope you find this information helpful. Of course you're free to decide on your own versioning. :upside_down_face: 

I'll let a maintainer help you with the Renovate technical parts, or configuration. :wink:

**Reply by paololazzari** [reply]
@HonkingGoose 

> Appending commits to the end of versions is harder to understand and use than plain versions.

I find it quite useful actually

**Reply by rarkins** [reply]
I'd recommend you adopt an approach like x.y.z-yyyymmddhhmmss-commitSha so that they sort correctly

**Reply by HonkingGoose** [reply]
For the latest Renovate release, it would look something like this:

`37.409.1-20240616160000-99cc62fa3ca89bf9de83a41a46c41d1525a46548`

That is a _very_ long string! At least the version information is up front. :upside_down_face:
